### ISSUE TITLE
Bear doesn't work with Bazel

### ISSUE THREAD

**Issue by martinlester** [rizsotto/Bear#170]
The readme for Bear requests notification of any build tools that use LD_PRELOAD, so here it is:

Bear doesn't work with Bazel, as it deliberately clears LD_PRELOAD and doesn't pass BEAR_OUTPUT (or other environment variables) to ensure hermeticity of builds.

As a hack around this, you can create a script "bazel-in-bear" in your path like this:

---

#!/bin/sh
cat <<END > ~/bin/gcc-in-bazel-in-bear
#!/bin/sh
env LD_PRELOAD=$LD_PRELOAD BEAR_OUTPUT=$BEAR_OUTPUT /usr/bin/env /usr/bin/gcc "\$@"
END
chmod u+x ~/bin/gcc-in-bazel-in-bear
env CC=gcc-in-bazel-in-bear bazel "$@"

---

This assumes ~/bin/ is in your path and your compiler is gcc. Launch the whole thing as:

bear bazel-in-bear build ...

Basically, it's creating another script that sets the environment variables appropriately and launches gcc, then using the CC environment variable to tell Bazel to use the script instead of calling gcc directly.

Not pretty, but it seems to work. I suspect use of Bear is not really compatible with Bazel's philosophy. There are forum posts about people adding rules to Bazel's BUILD to generate compile_commands, but their examples didn't work for me.

**Comment by rizsotto** [comment]
Thanks for the report. Was not tested Bear against Bazel. Will try it next week, till then I'm away from keyboard.

**Comment by vincent-picaud** [comment]
@martinlester I confirm the problem with Bear+Bazel, that is the reason why I have written [these scripts](https://github.com/vincent-picaud/Bazel_and_CompileCommands). 
@rizsotto  Maybe a little bit off topic but I take the opportunity to thank you for Bear, I use it with regular Makefiles, it works like a charm and is really helpful.

**Comment by rizsotto** [comment]
Thank you guys for your comments!
Took a quick look on the issue, and found you are right. Bear currently can't do much with daemon like build tools like Bazel. (The problem is not the `LD_PRELOAD` but the fact, that the compiler is not executed by the process we call rather than the daemon.) Maybe compiler wrappers could help, if I add to Bear later on.
Will update the README file known limitations section. Thanks again!

**Comment by Godin** [comment]
FWIW "--batch" option can be used to disable client/server mode.

Quoting https://bazel.build/versions/master/docs/bazel-user-manual.html :
> By default, Bazel uses a long-running server process as an optimization; this behavior can be disabled using the --batch option. There's no hard and fast rule about whether or not your script should use a server, but in general, the trade-off is between performance and reliability. **The server mode** makes a sequence of builds, especially incremental builds, faster, but its behavior **is more complex and prone to failure. We recommend in most cases that you use batch mode** unless the performance advantage is critical.

I assume that there is already no performance advantage as capture of all compilations requires "clean". 
HTH.

**Comment by damienmg** [comment]
Hi,
you can use `--action_env LD_PRELOAD` to ask bazel to pass around the LD_PRELOAD env variable.,

**Comment by rizsotto** [comment]
Thanks @damienmg , will give it a go again. Probably next week.

**Comment by rizsotto** [comment]
Was using and learned about bazel in the last 12 months, and concluded that Bear won't cover this build tool in the near future. (Will add a note about it in the readme file.)

**Comment by ak-ambi** [comment]
Was playing around with bear & bazel and found it possible to generate databse in similar way to martinlester:
```bash
bazel clean

bear -l /opt/lib/bear/libear.so bazel-in-bear --batch build --spawn_strategy=local ...
```

bazel-in-bear file:
```bash
#!/bin/sh
LIB=lib/x86_64-linux-gnu # Debian-specific
export LD_PRELOAD=$(eval "echo ${LD_PRELOAD}")
bazel "$@" --action_env=LD_PRELOAD=${LD_PRELOAD} --action_env=BEAR_OUTPUT=${BEAR_OUTPUT}
```

```--spawn_strategy local``` helped me to overcome the problem that bear is checking if source file exists before adding it to the compilation database. In sandboxed environment (default on Linux & macOS) all source files were filtered out.

@rizsotto I think it is possible to support Bazel, but probably some additional flags would be needed to support sandboxed compilation. I am thinking about the flag to configure mapping between path during sandboxed compilation and local path. This flag could also be useful when compiler is run from chrooted environment.

**Comment by cpsauer** [comment]
If anyone else lands here looking to generate compile_commands.json via Bazel...

We'd come here a while seeking ago the same, but ended up building https://github.com/hedronvision/bazel-compile-commands-extractor -- and just released it.

It makes things easy by directly asking Bazel for the list of commands it would run in a build, and thus doesn't have to fight Bazel's hermeticity strategy like BEAR does.
